File: sdk/src/main/java/com/cxense/cxensesdk/model/CustomParameter.java
package com.cxense.cxensesdk.model;

import androidx.annotation.NonNull;

import com.cxense.cxensesdk.Preconditions;
import com.google.gson.annotations.SerializedName;

/**
 * Customer-defined parameter object
 *
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2017-06-05).
 */

public final class CustomParameter {
    public static final String GROUP = "group";
    public static final String ITEM = "item";
    private static final String TYPE = "type";
    @SerializedName(GROUP)
    String name;
    @SerializedName(TYPE)
    String type;
    @SerializedName(ITEM)
    String item;

    private CustomParameter() {
    }

    /**
     * Create new instance of customer-defined parameter
     *
     * @param name Parameter name, e.g., "campaign", "adspace" or "creative".
     * @param item Parameter value, e.g. "sale", "42".
     */
    public CustomParameter(@NonNull String name, @NonNull String item) {
        Preconditions.checkStringForNullOrEmpty(name, "name");
        Preconditions.checkStringForNullOrEmpty(item, "item");
        this.name = name;
        this.item = item;
    }

    /**
     * Gets parameter name
     *
     * @return parameter name
     */
    @NonNull
    public String getName() {
        return name;
    }

    /**
     * Gets parameter value
     *
     * @return parameter value
     */
    @NonNull
    public String getItem() {
        return item;
    }
}

File: sdk/src/main/java/com/cxense/cxensesdk/model/CustomParameter.kt
package com.cxense.cxensesdk.model

import com.cxense.cxensesdk.model.CustomParameter.Companion.MAX_CUSTOM_PARAMETER_KEY_LENGTH
import com.cxense.cxensesdk.model.CustomParameter.Companion.MAX_CUSTOM_PARAMETER_VALUE_LENGTH
import com.google.gson.annotations.SerializedName

/**
 * Customer-defined parameter object
 * @param name Parameter name, e.g., "campaign", "adspace" or "creative".
 * @property name Parameter name, e.g., "campaign", "adspace" or "creative".
 *
 * @param value Parameter value, e.g. "sale", "42".
 * @property value Parameter value, e.g. "sale", "42".
 *
 * @throws IllegalArgumentException if [name] longer than [MAX_CUSTOM_PARAMETER_KEY_LENGTH] or [value] longer than [MAX_CUSTOM_PARAMETER_VALUE_LENGTH]
 */
@Suppress("unused", "MemberVisibilityCanBePrivate") // Public API.
class CustomParameter(
    name: String,
    value: String
) {
    @SerializedName(GROUP)
    val name: String = name.also {
        require(it.length <= MAX_CUSTOM_PARAMETER_KEY_LENGTH) {
            "Name can't be longer than $MAX_CUSTOM_PARAMETER_KEY_LENGTH symbols"
        }
    }
    @SerializedName(ITEM)
    val value: String = value.also {
        require(it.length <= MAX_CUSTOM_PARAMETER_VALUE_LENGTH) {
            "Value can't be longer than $MAX_CUSTOM_PARAMETER_VALUE_LENGTH symbols"
        }
    }

    companion object {
        const val GROUP = "group"
        const val ITEM = "item"
        const val MAX_CUSTOM_PARAMETER_KEY_LENGTH = 20
        const val MAX_CUSTOM_PARAMETER_VALUE_LENGTH = 256
    }
}
